[java] ForLoopCanBeForeach NPE when looping on `this` object

**Affects PMD Version:** 6.0.0

**Rule:** ForLoopCanBeForeach

**Description:** The rule should report loops that are iterated on `this` if possible. For now it considers neither `this.iterator()` nor just `iterator()`. Using `this.iterator()` like in the following code sample yields a NPE with the following stacktrace:
```scala
Caused by: java.lang.NullPointerException
	at net.sourceforge.pmd.lang.java.rule.bestpractices.ForLoopCanBeForeachRule.getIterableDeclOfIteratorLoop(ForLoopCanBeForeachRule.java:244)
	at net.sourceforge.pmd.lang.java.rule.bestpractices.ForLoopCanBeForeachRule.visit(ForLoopCanBeForeachRule.java:60)
	at net.sourceforge.pmd.lang.java.ast.ASTForStatement.jjtAccept(ASTForStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:424)
	at net.sourceforge.pmd.lang.java.ast.ASTStatement.jjtAccept(ASTStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:436)
	at net.sourceforge.pmd.lang.java.ast.ASTBlockStatement.jjtAccept(ASTBlockStatement.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:432)
	at net.sourceforge.pmd.lang.java.ast.ASTBlock.jjtAccept(ASTBlock.java:25)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:256)
	at net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration.jjtAccept(ASTMethodDeclaration.java:31)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:120)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBodyDeclaration.jjtAccept(ASTClassOrInterfaceBodyDeclaration.java:41)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:116)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceBody.jjtAccept(ASTClassOrInterfaceBody.java:21)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:92)
	at net.sourceforge.pmd.lang.java.ast.ASTClassOrInterfaceDeclaration.jjtAccept(ASTClassOrInterfaceDeclaration.java:33)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:232)
	at net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration.jjtAccept(ASTTypeDeclaration.java:35)
	at net.sourceforge.pmd.lang.java.ast.AbstractJavaNode.childrenAccept(AbstractJavaNode.java:56)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:83)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visit(AbstractJavaRule.java:212)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.visitAll(AbstractJavaRule.java:39)
	at net.sourceforge.pmd.lang.java.rule.AbstractJavaRule.apply(AbstractJavaRule.java:27)
	at net.sourceforge.pmd.RuleSet.apply(RuleSet.java:502)
	at net.sourceforge.pmd.RuleSets.apply(RuleSets.java:143)
	at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:181)
	at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:95)
	... 20 more
```
**Code Sample demonstrating the issue:**

```java
import java.util.Iterator;
class Foo<T> implements Iterable<T> {

  @Override
  public Iterator<T> iterator() {
    return null;
  }

  private void fofo() {
    for (Iterator<T> it = iterator(); it.hasNext();) {
      T item = it.next();
    }
    for (Iterator<T> it = this.iterator(); it.hasNext();) {
      T item = it.next();
    }
  }
}
```

**Running PMD through:** CLI

--- a/docs/pages/release_notes.md
+++ b/docs/pages/release_notes.md
@@ -30,6 +30,8 @@ This is a bug fixing release.
 *   java
     *   [#783](https://github.com/pmd/pmd/issues/783): \[java] GuardLogStatement regression
     *   [#793](https://github.com/pmd/pmd/issues/793): \[java] Parser error with private method in nested classes in interfaces
+*   java-bestpractices
+    *   [#800](https://github.com/pmd/pmd/issues/800): \[java] ForLoopCanBeForeach NPE when looping on `this` object
 *   java-design
     *   [#785](https://github.com/pmd/pmd/issues/785): \[java] NPE in DataClass rule
 

--- a/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/ForLoopCanBeForeachRule.java
+++ b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/ForLoopCanBeForeachRule.java
@@ -35,6 +35,10 @@ import net.sourceforge.pmd.lang.symboltable.Scope;
  */
 public class ForLoopCanBeForeachRule extends AbstractJavaRule {
 
+    public ForLoopCanBeForeachRule() {
+        addRuleChainVisit(ASTForStatement.class);
+    }
+    
     @Override
     public Object visit(ASTForStatement node, Object data) {
 
@@ -43,13 +47,13 @@ public class ForLoopCanBeForeachRule extends AbstractJavaRule {
         final ASTExpression guardCondition = node.getFirstChildOfType(ASTExpression.class);
 
         if (init == null && update == null || guardCondition == null) {
-            return super.visit(node, data);
+            return data;
         }
 
         Entry<VariableNameDeclaration, List<NameOccurrence>> indexDecl = getIndexVarDeclaration(init, update);
 
         if (indexDecl == null) {
-            return super.visit(node, data);
+            return data;
         }
 
 
@@ -67,7 +71,7 @@ public class ForLoopCanBeForeachRule extends AbstractJavaRule {
 
 
         if (occurrences == null || !"int".equals(index.getTypeImage()) || !indexStartsAtZero(index)) {
-            return super.visit(node, data);
+            return data;
         }
 
 
@@ -76,14 +80,14 @@ public class ForLoopCanBeForeachRule extends AbstractJavaRule {
 
 
         if (!isForUpdateSimpleEnough(update, itName) || iterableName == null) {
-            return super.visit(node, data);
+            return data;
         }
 
         Entry<VariableNameDeclaration, List<NameOccurrence>> iterableInfo = findDeclaration(iterableName, node.getScope());
         VariableNameDeclaration iterableDeclaration = iterableInfo == null ? null : iterableInfo.getKey();
 
         if (iterableDeclaration == null) {
-            return super.visit(node, data);
+            return data;
         }
 
         if (iterableDeclaration.isArray() && isReplaceableArrayLoop(node, occurrences, iterableDeclaration)) {
@@ -94,7 +98,7 @@ public class ForLoopCanBeForeachRule extends AbstractJavaRule {
             addViolation(data, node);
         }
 
-        return super.visit(node, data);
+        return data;
     }
 
 
@@ -240,8 +244,13 @@ public class ForLoopCanBeForeachRule extends AbstractJavaRule {
             return null;
         }
 
-        String name = initializer.getFirstDescendantOfType(ASTName.class)
-                                 .getImage();
+        ASTName nameNode = initializer.getFirstDescendantOfType(ASTName.class);
+        if (nameNode == null) {
+            // TODO : This can happen if we are calling a local / statically imported method that returns the iterable - currently unhandled
+            return null;
+        }
+        
+        String name = nameNode.getImage();
         int dotIndex = name.indexOf('.');
 
         if (dotIndex > 0) {

--- a/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/ForLoopCanBeForeach.xml
+++ b/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/ForLoopCanBeForeach.xml
@@ -266,5 +266,25 @@
         ]]></code>
     </test-code>
 
+	<test-code>
+        <description>Iterating on this object NPE, refs #800</description>
+        <expected-problems>0</expected-problems> <!-- FIXME should be 1 -->
+        <code><![CDATA[
+        import java.util.Iterator;
 
+        class Foo<T> implements Iterable<T> {
+
+            @Override
+            public Iterator<T> iterator() {
+                return null;
+            }
+
+            private void fofo() {
+                for (Iterator<T> it = this.iterator(); it.hasNext();) {
+                    T item = it.next();
+                }
+            }
+        }
+        ]]></code>
+    </test-code>
 </test-data>

